From 15bc1e44c4941ef6ce28a9b9cf3277e1915e18f4 Mon Sep 17 00:00:00 2001
From: gstavi <gur.stavi@gmail.com>
Date: Wed, 15 Jul 2015 15:58:55 +0300
Subject: [PATCH 02/10] better handling for matching braces
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------2.5.3"

This is a multi-part message in MIME format.
--------------2.5.3
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


diff --git a/PowerEditor/src/Notepad_plus.cpp b/PowerEditor/src/Notepad_plus.cpp
index d1ba610..79d290e 100644
--- a/PowerEditor/src/Notepad_plus.cpp
+++ b/PowerEditor/src/Notepad_plus.cpp
@@ -2046,29 +2046,27 @@ void Notepad_plus::findMatchingBracePos(int & braceAtCaret, int & braceOpposite)
 	int caretPos = int(_pEditView->execute(SCI_GETCURRENTPOS));
 	braceAtCaret = -1;
 	braceOpposite = -1;
-	TCHAR charBefore = '\0';
+	TCHAR BraceChar;
 
-	int lengthDoc = int(_pEditView->execute(SCI_GETLENGTH));
+    /* Gur: better handling for brace matching so repeat brace matching will
+     * always go back and forth */
 
-	if ((lengthDoc > 0) && (caretPos > 0))
-    {
-		charBefore = TCHAR(_pEditView->execute(SCI_GETCHARAT, caretPos - 1, 0));
+	int lengthDoc = int(_pEditView->execute(SCI_GETLENGTH));
+	/* Check if there is a brace after the caret */
+	if (lengthDoc > 0)
+	{
+		BraceChar = TCHAR(_pEditView->execute(SCI_GETCHARAT, caretPos, 0));
+		if (BraceChar && generic_strchr(TEXT("[](){}"), BraceChar))
+			braceAtCaret = caretPos;
 	}
-	// Priority goes to character before caret
-	if (charBefore && generic_strchr(TEXT("[](){}"), charBefore))
-    {
-		braceAtCaret = caretPos - 1;
+	/* No brace after, check if there is a brace before the caret */
+	if (braceAtCaret == -1 && (lengthDoc > 0) && (caretPos > 0))
+	{
+		BraceChar = TCHAR(_pEditView->execute(SCI_GETCHARAT, caretPos - 1, 0));
+		if (BraceChar && generic_strchr(TEXT("[](){}"), BraceChar))
+			braceAtCaret = caretPos - 1;
 	}
 
-	if (lengthDoc > 0  && (braceAtCaret < 0))
-    {
-		// No brace found so check other side
-		TCHAR charAfter = TCHAR(_pEditView->execute(SCI_GETCHARAT, caretPos, 0));
-		if (charAfter && generic_strchr(TEXT("[](){}"), charAfter))
-        {
-			braceAtCaret = caretPos;
-		}
-	}
 	if (braceAtCaret >= 0)
 		braceOpposite = int(_pEditView->execute(SCI_BRACEMATCH, braceAtCaret, 0));
 }

--------------2.5.3--


