From 80d962c098f17fba7ac2450cde8a1642155e11e7 Mon Sep 17 00:00:00 2001
From: gstavi <gur.stavi@gmail.com>
Date: Wed, 15 Jul 2015 16:23:38 +0300
Subject: [PATCH 05/11] Use MRU when choosing new file to activate after close
 file
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------2.5.3"

This is a multi-part message in MIME format.
--------------2.5.3
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


diff --git a/PowerEditor/src/Notepad_plus.cpp b/PowerEditor/src/Notepad_plus.cpp
index dc9ba79..4fe59d6 100644
--- a/PowerEditor/src/Notepad_plus.cpp
+++ b/PowerEditor/src/Notepad_plus.cpp
@@ -3159,6 +3159,27 @@ bool Notepad_plus::removeBufferFromView(BufferID id, int whichOne)
 			{
 				toActivate = active;    //activate the 'active' index. Since we remove the tab first, the indices shift (on the right side)
 			}
+
+			/*
+			 * Gur: After closing a file choose the file to activate (toActivate)
+			 * based on MRU list and not just last file in the list.
+			 */
+			if (NppParameters::getInstance()->getNppGUI()._styleMRU)
+			{
+				TaskListInfo taskListInfo;
+				::SendMessage(_pPublicInterface->getHSelf(), WM_GETTASKLISTINFO,
+					(WPARAM)&taskListInfo, 0);
+				int i, n = taskListInfo._tlfsLst.size();
+				for (i = 0; i < n ; i++)
+				{
+					TaskLstFnStatus &F = taskListInfo._tlfsLst[i];
+					if (F._iView != whichOne || F._bufID == (void *)id)
+						continue;
+					toActivate = F._docIndex >= active ? F._docIndex - 1 : F._docIndex;
+					break;
+				}
+			}
+
 			tabToClose->deletItemAt((size_t)index); //delete first
 			activateBuffer(tabToClose->getBufferByIndex(toActivate), whichOne);     //then activate. The prevent jumpy tab behaviour
 		}

--------------2.5.3--


