From 34a5334a39fbb772766a46e9de709aa7939d70b1 Mon Sep 17 00:00:00 2001
From: gstavi <gur.stavi@gmail.com>
Date: Wed, 15 Jul 2015 16:27:50 +0300
Subject: [PATCH 06/17] speedup test for modified files
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------2.30.0"

This is a multi-part message in MIME format.
--------------2.30.0
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


diff --git a/PowerEditor/src/Parameters.cpp b/PowerEditor/src/Parameters.cpp
index 5c266f2e..f4f6344f 100644
--- a/PowerEditor/src/Parameters.cpp
+++ b/PowerEditor/src/Parameters.cpp
@@ -6904,7 +6904,25 @@ void NppParameters::addScintillaModifiedIndex(int index)
 
 void NppParameters::safeWow64EnableWow64FsRedirection(BOOL Wow64FsEnableRedirection)
 {
-	HMODULE kernel = GetModuleHandle(TEXT("kernel32"));
+    typedef BOOL (WINAPI *LPFN_WOW64ENABLEWOW64FSREDIRECTION)(BOOL);
+    static LPFN_WOW64ENABLEWOW64FSREDIRECTION Wow64EnableWow64FsRedirectionFunc = NULL;
+    static int FirstTime = 1;
+    HMODULE kernel;
+
+	/*
+	 * Gur: speed optimization so we don't lookup kernel32 again and again.
+	 * This is called during the process of checking if files were modified
+	 * outside Notepad++.
+	 */
+	 if (!FirstTime)
+	 {
+		if (Wow64EnableWow64FsRedirectionFunc != NULL)
+			Wow64EnableWow64FsRedirectionFunc(Wow64FsEnableRedirection);
+		return;
+	}
+
+	FirstTime = 0;
+	kernel = GetModuleHandle(TEXT("kernel32"));
 	if (kernel)
 	{
 		BOOL isWow64 = FALSE;
@@ -6917,8 +6935,7 @@ void NppParameters::safeWow64EnableWow64FsRedirection(BOOL Wow64FsEnableRedirect
 
 			if (isWow64)
 			{
-				typedef BOOL (WINAPI *LPFN_WOW64ENABLEWOW64FSREDIRECTION)(BOOL);
-				LPFN_WOW64ENABLEWOW64FSREDIRECTION Wow64EnableWow64FsRedirectionFunc = (LPFN_WOW64ENABLEWOW64FSREDIRECTION)GetProcAddress(kernel, "Wow64EnableWow64FsRedirection");
+				Wow64EnableWow64FsRedirectionFunc = (LPFN_WOW64ENABLEWOW64FSREDIRECTION)GetProcAddress(kernel, "Wow64EnableWow64FsRedirection");
 				if (Wow64EnableWow64FsRedirectionFunc)
 				{
 					Wow64EnableWow64FsRedirectionFunc(Wow64FsEnableRedirection);
diff --git a/PowerEditor/src/ScitillaComponent/Buffer.cpp b/PowerEditor/src/ScitillaComponent/Buffer.cpp
index f257a700..249b67f0 100644
--- a/PowerEditor/src/ScitillaComponent/Buffer.cpp
+++ b/PowerEditor/src/ScitillaComponent/Buffer.cpp
@@ -221,13 +221,24 @@ void Buffer::setFileName(const TCHAR *fn, LangType defaultLang)
 
 bool Buffer::checkFileState() // returns true if the status has been changed (it can change into DOC_REGULAR too). false otherwise
 {
+	WIN32_FILE_ATTRIBUTE_DATA attributes;
+
+	/* Gur: speed optimization.
+	 * Optimistic approach - try to finish as quickly as possible */
+	if (_currentStatus == DOC_REGULAR)
+	{
+		BOOL res;
+		res = GetFileAttributesEx(_fullPathName.c_str(), GetFileExInfoStandard, &attributes);
+		if (res != 0 && CompareFileTime(&_timeStamp, &attributes.ftLastWriteTime) == 0)
+			return false;
+	}
+
 	// 1. Unsaved document cannot change by environment
 	// 2. Monitoring is sent by NPPM_INTERNAL_RELOADSCROLLTOEND
 	// So no need to check file status for both the above cases
 	if (_currentStatus == DOC_UNNAMED || isMonitoringOn())
 		return false;
 
-	WIN32_FILE_ATTRIBUTE_DATA attributes;
 	bool isWow64Off = false;
 	NppParameters& nppParam = NppParameters::getInstance();
 
@@ -514,6 +525,9 @@ void FileManager::checkFilesystemChanges(bool bCheckOnlyCurrentBuffer)
 {
 	if (bCheckOnlyCurrentBuffer)
 	{
+		if (_pNotepadPlus == NULL) {
+			return;
+		}
 		Buffer* buffer = _pNotepadPlus->getCurrentBuffer();
 		buffer->checkFileState();
 	}

--------------2.30.0--


