From 22bc067bc1cd22c01aae323c7fb1654313451f0d Mon Sep 17 00:00:00 2001
From: gstavi <gstavi@users.noreply.github.com>
Date: Thu, 7 Jul 2016 10:30:45 +0300
Subject: [PATCH 12/17] larger fonts
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------2.30.0"

This is a multi-part message in MIME format.
--------------2.30.0
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


diff --git a/PowerEditor/src/WinControls/FunctionList/functionListPanel.rc b/PowerEditor/src/WinControls/FunctionList/functionListPanel.rc
index a5c83308..5d2cb24a 100644
--- a/PowerEditor/src/WinControls/FunctionList/functionListPanel.rc
+++ b/PowerEditor/src/WinControls/FunctionList/functionListPanel.rc
@@ -33,7 +33,7 @@ IDD_FUNCLIST_PANEL DIALOGEX 26, 41, 142, 324
 STYLE DS_SETFONT | WS_POPUP | WS_CAPTION | WS_SYSMENU
 EXSTYLE WS_EX_TOOLWINDOW | WS_EX_WINDOWEDGE
 CAPTION "Function List"
-FONT 8, "MS Sans Serif", 0, 0, 0x0
+FONT 10, "MS Sans Serif", 0, 0, 0x0
 BEGIN
     //PUSHBUTTON      "Reload",IDC_RELOADBUTTON_FUNCLIST, 3,3,38,14
 	//LISTBOX         IDC_LIST_FUNCLIST,50,44,78,120,LBS_NOINTEGRALHEIGHT | WS_VSCROLL | WS_TABSTOP
diff --git a/PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.rc b/PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.rc
index de9488de..f3785e37 100644
--- a/PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.rc
+++ b/PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.rc
@@ -33,7 +33,7 @@ IDD_PROJECTPANEL DIALOGEX 26, 41, 142, 324
 STYLE DS_SETFONT | WS_POPUP | WS_CAPTION | WS_SYSMENU
 EXSTYLE WS_EX_TOOLWINDOW | WS_EX_WINDOWEDGE
 CAPTION "Project"
-FONT 8, "MS Sans Serif", 0, 0, 0x0
+FONT 10, "MS Sans Serif", 0, 0, 0x0
 BEGIN
 
 END
@@ -42,7 +42,7 @@ IDD_FILERELOCALIZER_DIALOG DIALOGEX 0, 0, 350, 48
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | WS_POPUP | WS_CAPTION | WS_SYSMENU
 EXSTYLE WS_EX_TOOLWINDOW
 CAPTION "Change file full path name"
-FONT 8, "MS Shell Dlg", 400, 0, 0x1
+FONT 10, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
     DEFPUSHBUTTON   "OK",IDOK,235,27,50,14
     PUSHBUTTON      "Cancel",IDCANCEL,290,27,50,14

--------------2.30.0--


